internal/server/iigointernal/executive.go

```diff
@@ -105,58 +128,64 @@ func (e *executive) replyAllocationRequest(commonPool shared.Resources) {
-	e.budget -= serviceCharge
+// appointing new role should be free now
+func (e *executive) appointNextSpeaker(clientIDs []shared.ClientID) (shared.ClientID, error) {
+	if !e.incurServiceCharge(actionCost.AppointNextSpeakerActionCost) {
+		return e.ID, errors.Errorf("Insufficient Budget in common Pool: appointNextSpeaker")
+	}
 	var election voting.Election
 	election.ProposeElection(baseclient.Speaker, voting.Plurality)
 	election.OpenBallot(clientIDs)
 	election.Vote(iigoClients)
-	return election.CloseBallot()
+	return election.CloseBallot(), nil
 }
 
-// withdrawSpeakerSalary withdraws the salary for speaker from the common pool.
-func (e *executive) withdrawSpeakerSalary(gameState *gamestate.GameState) bool {
-	var speakerSalary = shared.Resources(rules.VariableMap[rules.SpeakerSalary].Values[0])
-	var withdrawnAmount, withdrawSuccesful = WithdrawFromCommonPool(speakerSalary, gameState)
-	e.speakerSalary = withdrawnAmount
-	return withdrawSuccesful
-}
-
-// sendSpeakerSalary send speaker's salary to the speaker.
-func (e *executive) sendSpeakerSalary(legislativeBranch *legislature) {
+// sendSpeakerSalary conduct the transaction based on amount from client implementation
+func (e *executive) sendSpeakerSalary() error {
 	if e.clientPresident != nil {
-		amount, salaryPaid := e.clientPresident.PaySpeaker(e.speakerSalary)
-		if salaryPaid {
-			legislativeBranch.budget = amount
+		amount, speakerPaid := e.clientPresident.PaySpeaker()
+		if speakerPaid {
+			// Subtract from common resources pool
+			amountWithdraw, withdrawSuccess := WithdrawFromCommonPool(amount, e.gameState)
+
+			if withdrawSuccess {
+				// Pay into the client private resources pool
+				depositIntoClientPrivatePool(amountWithdraw, SpeakerIDGlobal, e.gameState)
+				return nil
+			}
 		}
-		return
 	}
-	legislativeBranch.budget = e.speakerSalary
+	return errors.Errorf("Cannot perform sendJudgeSalary")
 }
 
 func (e *executive) reset(val string) {
 	e.ID = 0
 	e.clientPresident = nil
-	e.budget = 0
 	e.ResourceRequests = map[shared.ClientID]shared.Resources{}
 	e.RulesProposals = []string{}
 	e.speakerSalary = 0
 }
 
-//requestRuleProposal asks each island alive for its rule proposal.
-func (e *executive) requestRuleProposal() {
-	e.budget -= serviceCharge
-	var rules []string
-	for _, island := range getIslandAlive() {
-		rules = append(rules, iigoClients[shared.ClientID(int(island))].RuleProposal())
-	}
-
-	e.setRuleProposals(rules)
+// Helper functions:
+func (e *executive) getTaxMap(islandsResources map[shared.ClientID]shared.Resources) (map[shared.ClientID]shared.Resources, bool) {
+	return e.clientPresident.SetTaxationAmount(islandsResources)
 }
 
 func getIslandAlive() []float64 {
 	return rules.VariableMap[rules.IslandsAlive].Values
 }
+
+// incur charges in both budget and commonpool for performing an actions
+// actionID is typically the function name of the action
+// only return error if we can't withdraw from common pool
```

Comment:
It doesn't seem to return an error here. Is this intentional?

Reply:
Think it should be fine. Basically, we return if it was successful or not. Thanks for checking